examples/pizza/web/index.js

```diff
@@ -0,0 +1,144 @@
+
+  }).then(function(characteristics) {
+    characteristicNames.forEach(function(name, i) {
+      cachedCharacteristics[name] = characteristics[i];
+    });
+    return cachedCharacteristics;
+
+  }).catch(function(err) {
+    alert('oven (bluetooth) error');
+    throw err;
+  });
+};
+
+// characteristic setup
+var readyCrust = function(crustType) {
+  console.log('crustType:', crustType);
+  return function() {
+    var crust = new Uint8Array(1);
+    crust[0] = crustType;
+
+    var pizzaCrustCharacteristic = cachedCharacteristics['crust'];
+    if(pizzaCrustCharacteristic == null) throw new Error('oven not ready!');
+    return pizzaCrustCharacteristic.writeValue(crust).catch(function(err) {
+      alert('crust error');
+      throw err;
+    });
+  }
+};
+
+var readyToppings = function(toppings) {
+  console.log('toppings:', toppings);
+  return function() {
+    var toppingsBuff = new Uint8Array(2);
+    toppingsBuff[0] = toppings.concat(0).reduce((a, b)=>a | b);
+
+    var pizzaToppingsCharacteristic = cachedCharacteristics['toppings'];
+    if(pizzaToppingsCharacteristic == null) throw new Error('oven not ready');
+    return pizzaToppingsCharacteristic.writeValue(toppingsBuff).catch(function(err) {
+      alert('toppings error');
+      throw err;
+    });
+  };
+};
+
+var bakePizza = function(temperature) {
+  console.log('oven temp:', temperature);
+  return function() {
+    var pizzaBakeCharacteristic = cachedCharacteristics['bake'];
+    if(pizzaBakeCharacteristic == null) throw new Error('oven not ready!');
+
+    var tempBuff = new Uint16Array([swap16(temperature)]);
+    return pizzaBakeCharacteristic.writeValue(tempBuff).then(function() {
+      result = pizzaBakeCharacteristic.value.getUint8();
+      log('The result is ' + (
+        result == PizzaBakeResult.HALF_BAKED ? 'half baked.' :
+        result == PizzaBakeResult.BAKED ? 'baked.' :
+        result == PizzaBakeResult.CRISPY ? 'crispy.' :
+        result == PizzaBakeResult.BURNT ? 'burnt.' :
+        result == PizzaBakeResult.ON_FIRE ? 'on fire!' : 'unknown?'));
+
+      return result;
+
+    }).catch(function(err) {
+      alert('bake error');
+      throw err;
+    });
+  };
+};
+
+// get values from dom
+var getCrustType = function() {
+  return Number(crustSelectEl.value);
+};
+
+var getToppings = function() {
+  var toppings = [];
+  [].slice.call(toppingsEls).forEach(function(el) {
+    if(el.checked) toppings.push(Number(el.value));
+  });
+  return toppings;
+};
+
+var getOvenTemperature = function() {
+  return ovenTempEl.value;
+};
+
+
+// button listeners
+var onStartButtonClick = function(e) {
+  if(ovenServer != null && ovenServer.connected) alert('Already connected...');
+  return readyOven().then(function() {
+    alert('Connection successful!');
+    console.log(ovenServer);
+  });
+};
+
+var onBakeButtonClick = function(e) {
+  if(ovenServer == null || !ovenServer.connected) alert('Not connected!');
+  [
```

Comment:
The issue I'm seeing with this code is that `bakePizza` will be called even if `readyToppings` rejects sadly.
How about a simpler version?

```js
Promise.resolve()
.then(_ => readyCrust(getCrustType()))
.then(_ => readyToppings(getToppings()))
.then(_ => bakePizza(getOvenTemperature()))